daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -31,86 +30,85 @@ import scala.reflect.ClassTag
   * using `Seq` encourages patterns where the performance of what you're doing is totally
   * unclear. Use `toSeq` if you want a `Seq`, and think about what that means.
   */
+// The class is backed by an array of objects. This means that elements of primitive types are boxed.
 final class ImmArray[+A] private (
     private val start: Int,
     val length: Int,
-    array: mutable.ArraySeq[A]
+    private val array: Array[AnyRef]
```

Comment:
I think trickery is called for here. The type `boxedArray.Boxed` becomes a substitute for `A`, with implicit conversions each way between them. Then you can get rid of most of the casts and be sure you aren't mixing it up with a different `AnyRef` (which is what `boxedArray.Boxed` really is).

```scala
  private[this] val boxedArray = AnyRefIs[A](_array)
  import boxedArray.{Boxed, aBoxed, aUnboxed}
  private def array: Array[_ <: AnyRef] = boxedArray.array

private sealed abstract class AnyRefIs[A] {
  type Boxed <: AnyRef
  val array: Array[Boxed]
  implicit def aBoxed(a: A): Boxed = a.asInstanceOf[Boxed]
  implicit def aUnboxed(boxed: Boxed): A = boxed.asInstanceOf[A]
}

private object AnyRefIs {
  def apply[A](_array: Array[_ <: AnyRef]): AnyRefIs[A] = new AnyRefIs[A] {
    type Boxed = AnyRef
    override val array = _array
  }
}
```

`AnyRefIs` can be inlined within `ImmArray` by making `ImmArray` abstract.